lib/page_ez/selector_evaluator.rb

```diff
@@ -0,0 +1,76 @@
+module PageEz
+  class SelectorEvaluator
+    def self.build(name, dynamic_options:, options:, selector:)
+      run_def = ->(args, target:) do
+        PageEz::SelectorEvaluator.new(name, args, dynamic_options: dynamic_options, options: options, selector: selector, target: target)
+      end
+
+      name_def = -> { name }
+
+      Class.new.tap do |klass|
+        klass.define_singleton_method(:run, &run_def)
+        klass.define_singleton_method(:name, &name_def)
+      end
+    end
+
+    def initialize(name, args, dynamic_options:, options:, selector:, target:)
+      @name = name
+      @args = args
+      @dynamic_options = dynamic_options
+      @options = options
+      @selector = if selector.respond_to?(:bind)
+        selector.bind(target)
+      else
+        selector
+      end
+    end
+
+    def selector
+      if dynamic_selector?
+        if selector_kwargs.any?
+          @selector.call(**kwargs.slice(*selector_kwargs))
+        else
+          @selector.call(*args[0..selector_args.length - 1])
+        end
```

Comment:
My understanding is that this allows either positional args or keyword args, _but not both_. I've found an implementation that appears to work with all the existing specs, and a new spec that accepts both a positional argument and a keyword argument.

```ruby
    test_page = Class.new(PageEz::Page) do
      # This probably isn't the best example, but it was an easy way to adjust
      # the existing spec
      has_one :element_with_id do
        has_one :heading
      end

      def element_with_id(element, id:)
        "#{element}[data-id='#{id}']"
      end
    end.new(page)

    page.visit "/"

    # This passes
    expect(test_page.element_with_id("section", id: 1)).to have_text("Section 1")
```

```suggestion
        if selector_args.none?
          @selector.call(**kwargs.slice(*selector_kwargs))
        else
          @selector.call(*args[0..selector_args.length - 1], **kwargs.slice(*selector_kwargs))
        end
```

I was unable to find a way to omit the conditional, because [Array#[]](https://ruby-doc.org/3.2.2/Array.html#method-i-3D-3D) notes:

> Special case: If range.start == a.size, returns a new empty Array.

This was causing an empty array to be passed as a positional argument when one wasn't expected.

Reply:
Great catch! I'd kept it in here but as a separate commit: https://github.com/joshuaclayton/page_ez/pull/30/commits/59507eff670ba01f8b5976eadd09f95d4bda140b Appreciate the code suggestion!